src/core/ApolloClient.ts

```diff
@@ -342,8 +342,8 @@ export class ApolloClient<TCacheShape> implements DataProxy {
    */
   public mutate<
     TData = any,
-    TVariables = OperationVariables,
-    TContext = DefaultContext,
+    TVariables extends OperationVariables = OperationVariables,
+    TContext extends {} = DefaultContext,
```

Comment:
Is this `{}` type meant to represent a type that is an object? The `{}` type actually means "any non-nullish value", not an object, so we probably want to be a bit more specific here.

You can read more about it in [this comment](https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-675156492) posted to ESLint, which has an [ESLint rule](https://typescript-eslint.io/rules/ban-types) around this particular behavior.

Instead, we should opt to check against `Record<string, any>` to ensure this is scoped to object types.

```suggestion
    TContext extends Record<string, any> = DefaultContext,
```

Reply:
Here is a great playground that demonstrates this oddity: https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdEBvAXwD4AKABwEMAnRgWwC5EMBKCygFACUcWlGYhsPANzDUtAIwy5oqspG0wIADbaZiAPQGcuetmjYiOZszjMBQA

Reply:
Thanks for the review, @jerelmiller! I'm just getting up to speed on the changes in TS >= `4.8` and I think this reference thread adds context around some of your comments/questions: https://github.com/microsoft/TypeScript/issues/49489

There are definitely opportunities to make some generic params narrower/more specific; the initial goal here was to address unconstrained parameters where the previous behavior would have allowed `null`/`undefined` to be passed ([docs](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-8.html#unconstrained-generics-no-longer-assignable-to-)) when `strictNullChecks` is `true`. For that reason, the docs suggest `{}` as an initial step to narrowing unconstrained parameters given that it's `undefined`/`null` that should be disallowed, but we can definitely use `T extends object` (or e.g. `Record<string, unknown>`) where strings, numbers, and booleans are also not expected.

Reply:
That makes sense. Thanks for the reference! I trust you made the right call in the right places.